src/cache.ts

```diff
@@ -0,0 +1,114 @@
+import { KubernetesObject } from './types';
+import { Watch } from './watch';
+
+export interface ObjectCache<T> {
+    get(name: string, namespace?: string): T | null;
+    list(namespace?: string): T[];
+}
+
+export type ListCallback<T extends KubernetesObject> = (list: T[]) => void;
+
+export class ListWatch<T extends KubernetesObject> implements ObjectCache<T> {
+    private objects: T[] = [];
+    private indexCache: any = {};
+    private path: string;
+    private watch: Watch;
+    private listFn: (callback: ListCallback<T>) => void;
+
+    public constructor(path: string, watch: Watch, listFn: (callback: ListCallback<T>) => void) {
+        this.watch = watch;
+        this.listFn = listFn;
+        this.path = path;
+        this.doneHandler(null);
+    }
+
+    public get(name: string, namespace?: string): T | null {
+        let result: T | null = null;
+        for (const element of this.objects) {
+            if (element.metadata.name === name &&
+                (!namespace || element.metadata.namespace === namespace)) {
+                result = element;
+            }
+        }
+        return result;
+    }
+
+    public list(namespace?: string | undefined): T[] {
+        if (!namespace) {
+            return this.objects;
```

Comment:
Should we make a copy here?  Are we worried if some does a `list(...).push(...)`?

Reply:
Switched to `ReadonlyArray<T>`

Reply:
HOW DID I NOT KNOW ABOUT `ReadonlyArray<T>` HEADS WILL ROLL FOR THIS.  Thank you, that's some TypeScript learning for me today too!